contracts/poap/src/contract.rs

```diff
@@ -0,0 +1,1375 @@
+            },
+            result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn event_info_end_time_before_current_time_error() {
+        let mut deps = mock_dependencies();
+        let mut env = mock_env();
+        let info = mock_info(CREATOR, &vec![]);
+
+        do_instantiate(deps.as_mut());
+        env.block.time = Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS);
+
+        let msg = ExecuteMsg::UpdateEventInfo {
+            start_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS + 2),
+            end_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS - 1),
+        };
+
+        let result = execute(deps.as_mut(), env.clone(), info.clone(), msg);
+        assert_eq!(
+            ContractError::StartTimeAfterEndTime {
+                start: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS + 2),
+                end: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS - 1),
+            },
+            result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn event_info_end_time_equal_current_time_error() {
+        let mut deps = mock_dependencies();
+        let mut env = mock_env();
+        let info = mock_info(CREATOR, &vec![]);
+
+        do_instantiate(deps.as_mut());
+        env.block.time = Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS);
+
+        let msg = ExecuteMsg::UpdateEventInfo {
+            start_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS + 2),
+            end_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS),
+        };
+
+        let result = execute(deps.as_mut(), env.clone(), info.clone(), msg);
+        assert_eq!(
+            ContractError::StartTimeAfterEndTime {
+                start: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS + 2),
+                end: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS),
+            },
+            result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn update_admin_only_from_admin() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        const NEW_ADMIN: &str = "admin2";
+
+        do_instantiate(deps.as_mut());
+
+        let info = mock_info(USER, &vec![]);
+        let msg = ExecuteMsg::UpdateAdmin {
+            new_admin: NEW_ADMIN.to_string(),
+        };
+
+        let result = execute(deps.as_mut(), env.clone(), info, msg.clone());
+        assert_eq!(ContractError::Unauthorized {}, result.unwrap_err());
+
+        let info = mock_info(CREATOR, &vec![]);
+
+        let result = execute(deps.as_mut(), env.clone(), info, msg.clone());
+        assert_eq!(ContractError::Unauthorized {}, result.unwrap_err());
+
+        let info = mock_info(ADMIN, &vec![]);
+
+        execute(deps.as_mut(), env.clone(), info, msg.clone()).unwrap();
+
+        let config = CONFIG.load(&deps.storage).unwrap();
+        assert_eq!(NEW_ADMIN, config.admin.as_str());
+    }
+
+    #[test]
+    fn update_minter_only_from_admin() {
```

Comment:
as above, it can be splited into:
1. update_minter_without_permission_error
2. update_minter_with_permission_properly